theano/scan_module/scan_opt.py

```diff
@@ -219,106 +228,123 @@ def apply(self, fgraph):
             self.process_node(fgraph, node)
 
     def process_node(self, fgraph, node):
+        """
+        IMPORTANT NOTE: This function uses set and dictionary data structures.
+        By default they are not ordered for efficiency reasons. Take care
+        and make sure of changing them with their Ordered counterparts if you
+        need to iterate over these variables.
+        """
         # this flag tells if there was any change during the last iterations
-        changed = True
         clean_inputs, clean_outputs = scan_utils.reconstruct_graph(
             node.op.inputs, node.op.outputs)
 
-        local_fgraph = gof.FunctionGraph(clean_inputs, clean_outputs, clone=False)
-        max_iterations = 2 * len(local_fgraph.toposort()) + 3
-        counts = 0
-        to_remove = []
-        to_replace = []
+        local_fgraph = gof.FunctionGraph(clean_inputs,
+                                         clean_outputs,
+                                         clone=False)
+
+        local_fgraph_topo = local_fgraph.toposort()
+        local_fgraph_outs_set = set(local_fgraph.outputs)
+        local_fgraph_outs_map = dict([(v, k) for k, v in \
+                enumerate(local_fgraph.outputs)])
+
+        to_remove_set = set()
+        to_remove_add = to_remove_set.add
```

Comment:
This shorthand for to_remove_set.add should probably be removed. It's only marginally shorter than directly calling to_remove_set.add() and it doesn't really improve readability.